pkg/merger/merger.go

```diff
@@ -0,0 +1,129 @@
+limitations under the License.
+*/
+
+package merger
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/golang/protobuf/proto"
+	"github.com/sirupsen/logrus"
+	"gopkg.in/yaml.v2"
+
+	"github.com/GoogleCloudPlatform/testgrid/config"
+	configpb "github.com/GoogleCloudPlatform/testgrid/pb/config"
+	"github.com/GoogleCloudPlatform/testgrid/util/gcs"
+)
+
+type configList struct {
+	Target  string    `json:"Target"`
+	Sources []sources `json:"Sources"`
+}
+
+type sources struct {
+	Name     string `json:"Name"`
+	Location string `json:"Location"`
+}
+
+// ParseAndCheck parses and checks the configuration file for common errors
+func ParseAndCheck(data []byte) (target *gcs.Path, sources map[string]*gcs.Path, err error) {
+	cfg := &configList{}
+	err = yaml.Unmarshal(data, cfg)
+	if err != nil {
+		return
+	}
+
+	target, err = gcs.NewPath(cfg.Target)
+	if err != nil {
+		return
+	}
+
+	sources = make(map[string]*gcs.Path)
+	for _, source := range cfg.Sources {
+		if _, exists := sources[source.Name]; exists {
+			return nil, nil, fmt.Errorf("duplicated name %s", source.Name)
+		}
+		path, err := gcs.NewPath(source.Location)
+		if err != nil {
+			return nil, nil, err
+		}
+		sources[source.Name] = path
+	}
+
+	if len(sources) == 0 {
+		return target, sources, errors.New("no shards to converge")
+	}
+	return
+}
+
+type mergeClient interface {
+	gcs.Opener
+	gcs.Uploader
+}
+
+// MergeAndUpdate gathers configurations from each path and merges them.
+// Puts the result at targetPath if confirm is true
+// Will skip an input config if it is invalid and skipValidate is false
+// Other problems are considered fatal and will return an error
+func MergeAndUpdate(ctx context.Context, client mergeClient, paths map[string]*gcs.Path, targetPath gcs.Path, skipValidate, confirm bool) error {
+	// Deserialize each proto
+	// TODO: Reading and validating can be done in parallel with a wait group.
+	// TODO: Cache the version for each source. Only read if they've changed.
+	shards := make(map[string]*configpb.Configuration)
+
+	for key, path := range paths {
+		if path == nil {
+			return fmt.Errorf("path at '%s' is nil", key)
+		}
+		cfg, err := config.ReadGCS(ctx, client, *path)
+		if err != nil {
+			return fmt.Errorf("can't read config %s at %v: %w", key, path, err)
+		}
+		if !skipValidate {
+			if err := config.Validate(cfg); err != nil {
+				logrus.WithError(err).Errorf("config %s at %v failed validation; skipping config", key, path)
+				continue
+			}
+		}
+		shards[key] = cfg
+	}
+
+	if len(shards) == 0 {
+		return errors.New("no configs to merge")
+	}
+
+	// Merge and marshal the result
+	result, err := config.Converge(shards)
+	if err != nil {
+		return fmt.Errorf("failed to merge configurations: %w", err)
+	}
+
+	buf, err := proto.Marshal(result)
+	if err != nil {
+		return fmt.Errorf("failed to serialize merged proto: %w", err)
+	}
+
+	if confirm {
+		err = client.Upload(ctx, targetPath, buf, gcs.DefaultAcl, "no-cache")
+		if err != nil {
```

Comment:
combine with previous line:

```golang
// since this is inside a {} you can also do if err := client.Upload()
if err = client.Upload(...); err != nil {
  return fmt.Errorf("could not...")
}
```